src/OpenTelemetry.Contrib.Instrumentation.Wcf/TelemetryDispatchMessageInspector.cs

```diff
@@ -0,0 +1,111 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+#if NETFRAMEWORK
+using System;
+using System.Diagnostics;
+using System.ServiceModel;
+using System.ServiceModel.Channels;
+using System.ServiceModel.Dispatcher;
+using OpenTelemetry.Contrib.Instrumentation.Wcf.Implementation;
+using OpenTelemetry.Trace;
+
+namespace OpenTelemetry.Contrib.Instrumentation.Wcf
+{
+    /// <summary>
+    /// An <see cref="IDispatchMessageInspector"/> implementation which adds telemetry to incoming requests.
+    /// </summary>
+    public class TelemetryDispatchMessageInspector : IDispatchMessageInspector
+    {
+        /// <inheritdoc/>
+        public object AfterReceiveRequest(ref Message request, IClientChannel channel, InstanceContext instanceContext)
+        {
+            try
+            {
+                if (WcfInstrumentationActivitySource.Options == null || WcfInstrumentationActivitySource.Options.IncomingRequestFilter?.Invoke(request) == false)
+                {
+                    WcfInstrumentationEventSource.Log.RequestIsFilteredOut(request.Headers.Action);
+                    return null;
+                }
+            }
+            catch (Exception ex)
+            {
+                WcfInstrumentationEventSource.Log.RequestFilterException(ex);
+                return null;
+            }
+
+            var ctx = WcfInstrumentationActivitySource.Options.Propagator.Extract(default, request, WcfInstrumentationActivitySource.MessageHeaderValuesGetter);
+
+            Activity activity = WcfInstrumentationActivitySource.ActivitySource.StartActivity(
+                WcfInstrumentationActivitySource.IncomingRequestActivityName,
+                ActivityKind.Server,
+                ctx.ActivityContext);
+
+            if (activity != null)
+            {
+                string action = request.Headers.Action;
+                activity.DisplayName = action;
+
+                if (activity.IsAllDataRequested)
+                {
+                    int lastIndex = action.LastIndexOf('/');
+
+                    activity.SetTag(WcfInstrumentationConstants.RpcSystemTag, WcfInstrumentationConstants.WcfSystemValue);
+                    activity.SetTag(WcfInstrumentationConstants.RpcServiceTag, action.Substring(0, lastIndex));
+                    activity.SetTag(WcfInstrumentationConstants.RpcMethodTag, action.Substring(lastIndex + 1));
+                    activity.SetTag(WcfInstrumentationConstants.NetHostNameTag, channel.LocalAddress.Uri.Host);
+                    activity.SetTag(WcfInstrumentationConstants.NetHostPortTag, channel.LocalAddress.Uri.Port);
+
+                    if (WcfInstrumentationActivitySource.Options.SetSoapVersion)
```

Comment:
Technically request.Version isn't the SOAP version, it's the MessageVersion which is a tuple of the EnvelopeVersion and AddressingVersion. EnvelopeVersion is the version of SOAP used with the message. I don't think you need to change any property names as this is semantics which only people like me see the significance of, but might be worth making sure any documentation is concise.

Reply:
Updated